src/librustc_typeck/check/writeback.rs

```diff
@@ -523,8 +532,25 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         self.visit_adjustments(span, hir_id);
 
         // Resolve the type of the node with id `node_id`
-        let n_ty = self.fcx.node_ty(hir_id);
-        let n_ty = self.resolve(&n_ty, &span);
+        let n_ty_original = self.fcx.node_ty(hir_id);
+        let n_ty = self.resolve(&n_ty_original, &span);
+
+        // XXX check whether the node type contains any of the type variables that
+        // became `!` as a result of type fallback. if so, warn.
+        if !self.from_diverging_fallback.is_empty() {
```

Comment:
basically here, we would check something like 

* if `hir_id` is in the "dead node" set
* and `n_ty_original` is in the "from diverging fallback" set

the code as it is now is actually testing more than I  think we should test, since it is checking whether the type *contains any type that fell back to `!`*. But this will also warn for a type like `Result<(), !>` where the `!` resulted from fallback. As we discussed, I think that's probably too broad, and we only want to warn if `n_ty_original` is literally inside of `self.from_diverging_fallback`, but we should experiment.

Reply:
As we said on Zulip, I think what I meant is 

* if `hir_id` is NOT in the dead node set (i.e., it is live)
* and `n_ty_original` is in the 'from diverging fallback' set

then issue a warning.

Reply:
Makes sense !